doc/source/development/developer.rst

```diff
@@ -53,26 +60,43 @@ for each column, *including the index columns*. This has JSON form:
     'numpy_type': numpy_type,
     'metadata': metadata}
 
-.. note::
+See below for the detailed specification for these
+
+Index Metadata Descriptors
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``RangeIndex`` can be stored as metadata only, not requiring serialization. The
+descriptor format for these as is follows:
+
+.. code-block:: python
+
+   index = pd.RangeIndex(0, 10, 2)
+   {'kind': 'range',
+    'name': index.name,
+    'start': index._start,
+    'stop': index._stop,
+    'step': index._step}
 
-   Every index column is stored with a name matching the pattern
-   ``__index_level_\d+__`` and its corresponding column information is can be
-   found with the following code snippet.
+Other index types must be serialized as data columns along with the other
+DataFrame columns. The metadata for these is a dict with ``kind`` field
+``'serialized'`` and ``'field_name'`` field indicating which data column
+contains the index data. For example,
 
-   Following this naming convention isn't strictly necessary, but strongly
-   suggested for compatibility with Arrow.
+.. code-block:: python
 
-   Here's an example of how the index metadata is structured in pyarrow:
+   {'kind': 'serialized',
+    'field_name': '__index_level_0__'}
 
-    .. code-block:: python
+Every index column is stored with a name matching the pattern
+``__index_level_\d+__``. Following this naming convention isn't strictly
+necessary, but strongly suggested for compatibility with Arrow and
+disambiguation. The ``'field_name'`` is the actual name of the column in the
+serialized Parquet table. If the ``Index`` has a non-None ``name`` attribute,
+then it can be found in the ``name`` field of the metadata for that serialized
+data column as described below.
```

Comment:
Is this actually correct? (although it was already there in the doc before, to be clear) 
This seems to indicate that an index always gets a ``__index_level_x__`` name as the `field_name`, regardless of the name it has (so not only if it is None).

But this is not what I see from a quick test:

```
In [3]: pyarrow.__version__                                                                                                                                                     
Out[3]: '0.12.0'

In [4]: df = pd.DataFrame({'a': [1, 2, 3], 'b': [3, 4, 5]}).set_index('a')                                                                                                      

In [5]: df                                                                                                                                                                      
Out[5]: 
   b
a   
1  3
2  4
3  5

In [6]: pyarrow.Table.from_pandas(df)                                                                                                                                           
Out[6]: 
pyarrow.Table
b: int64
a: int64
metadata
--------
OrderedDict([(b'pandas',
              b'{"index_columns": ["a"], "column_indexes": [{"name": null, "'
              b'field_name": null, "pandas_type": "unicode", "numpy_type": "'
              b'object", "metadata": {"encoding": "UTF-8"}}], "columns": [{"'
              b'name": "b", "field_name": "b", "pandas_type": "int64", "nump'
              b'y_type": "int64", "metadata": null}, {"name": "a", "field_na'
              b'me": "a", "pandas_type": "int64", "numpy_type": "int64", "me'
              b'tadata": null}], "pandas_version": "0.23.4"}')])
```

(I remember that we had a discussion about this before, but can't directly remember the outcome of that)

Reply:
It appears that the current behavior is to use the name of the index if it does not conflict with any of the data columns. So we should update these docs to reflect that